src/net_processing.cpp

```diff
@@ -2777,11 +2787,28 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         std::vector<CAddress> vAddrOk;
         int64_t nNow = GetAdjustedTime();
         int64_t nSince = nNow - 10 * 60;
+
+        // Update/increment addr rate limiting bucket.
+        const auto current_time = GetTime<std::chrono::microseconds>();
+        if (peer->m_addr_token_bucket < MAX_ADDR_TO_SEND) {
+            // Don't increment bucket if it's already full
+            auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
+            double increment = std::chrono::duration<double>(time_diff).count() * MAX_ADDR_RATE_PER_SECOND;
+            peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_TO_SEND);
+        }
+        peer->m_addr_token_timestamp = current_time;
+
+        Shuffle(vAddr.begin(), vAddr.end(), FastRandomContext());
```

Comment:
The idea is that if we're going to be dropping some part of incoming messages because of rate-limiting, we should do it uniformly. This isn't to stop any kind of attack, but to give slightly fairer treatment to existing nodes on the network, which don't randomize on sending (there is random replacement once the outgoing buffer hits 1000, but that is rarely if ever hit under normal network conditions).